os/fs/smartfs/smartfs_smart.c

```diff
@@ -712,24 +716,26 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer, size_t
 				goto errout_with_semaphore;
 			}
 
-			/* If file is modified and more data remains to be appended to file, but no next sector is available,
-			 * do not update sf->currsector and curroffset. Will be handled when buffer data is synced.
+			/* If file is modified and more data remains to be appended to file,
+			 * but no next sector is available, do not update sf->currsector and curroffset.
+			 * This will be handled when buffer data is synced.
 			 */
 			if (SMARTFS_NEXTSECTOR(header) != 0xFFFF) {
-				sf->currsector = SMARTFS_NEXTSECTOR(header);
-				
 				/* Now get the chained sector info and reset the offset */
+				sf->currsector = SMARTFS_NEXTSECTOR(header);
 				sf->curroffset = size;
+
 #ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-				/* Read the next sector if more bytes remain in the write buffer */
-				if (buflen > 0) {
```

Comment:
@Taejun-Kwon 
buflen passed to smartfs_write = fs->fs_llformat.availbytes
Hence, every time we write optimal buffer size, after that buflen = 0
So this part of code does not get executed and next sector is not read

It is fine when we are not waiting for smartfs_sync_internal, because Line 683-684 above do not write header
They just write file data

But when sync_internal writes last part of data, it needs the header which is not there in sf->buffer

Reply:
I think rollback #5105 is better option, we don't have to read whole of sector if buflen from high layer is not exist. only header is necessary

Reply:
Updated with an alternate solution, please take a look.